backend/cmd/api/loaders.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"github.com/bcc-code/brunstadtv/backend/common"
+	"github.com/bcc-code/brunstadtv/backend/items/collection"
+	"github.com/bcc-code/brunstadtv/backend/loaders"
+	"github.com/bcc-code/brunstadtv/backend/sqlc"
+	"github.com/google/uuid"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+)
+
+var roleLoaders = sync.Map{}
+
+func getLoadersForRoles(db *sql.DB, queries *sqlc.Queries, collectionLoader *loaders.Loader[int, *common.Collection], roles []string) *common.FilteredLoaders {
+	sort.Strings(roles)
+
+	key := strings.Join(roles, "-")
+
+	if ls, ok := roleLoaders.Load(key); ok {
+		return ls.(*common.FilteredLoaders)
+	}
+
+	ctx := context.Background()
+
+	rq := queries.RoleQueries(roles)
+
+	ls := &common.FilteredLoaders{
+		ShowFilterLoader:    loaders.NewFilterLoader(ctx, rq.GetShowIDsWithRoles),
+		SeasonFilterLoader:  loaders.NewFilterLoader(ctx, rq.GetSeasonIDsWithRoles),
+		EpisodeFilterLoader: loaders.NewFilterLoader(ctx, rq.GetEpisodeIDsWithRoles),
+		SeasonsLoader:       loaders.NewRelationLoader(ctx, rq.GetSeasonIDsForShowsWithRoles),
+		SectionsLoader:      loaders.NewRelationLoader(ctx, rq.GetSectionIDsForPagesWithRoles),
+		EpisodesLoader:      loaders.NewRelationLoader(ctx, rq.GetEpisodeIDsForSeasonsWithRoles),
+		CollectionItemsLoader: loaders.NewListLoader(ctx, rq.GetItemsForCollectionsWithRoles, func(i common.CollectionItem) int {
+			return i.CollectionID
+		}),
+		CollectionItemIDsLoader: collection.NewCollectionItemLoader(ctx, db, collectionLoader, roles),
+		CalendarEntryLoader:     loaders.New(ctx, rq.GetCalendarEntries),
+		StudyTopicFilterLoader:  loaders.NewFilterLoader(ctx, rq.GetTopicIDsWithRoles),
+		StudyLessonFilterLoader: loaders.NewFilterLoader(ctx, rq.GetLessonIDsWithRoles),
+		StudyTaskFilterLoader:   loaders.NewFilterLoader(ctx, rq.GetTaskIDsWithRoles),
+		StudyLessonsLoader:      loaders.NewRelationLoader(ctx, rq.GetLessonIDsForTopics),
+		StudyTasksLoader:        loaders.NewRelationLoader(ctx, rq.GetTaskIDsForLessons),
+
+		// Study Relations
+		StudyLessonEpisodesLoader: loaders.NewRelationLoader(ctx, rq.GetEpisodeIDsForLessons),
+		EpisodeStudyLessonsLoader: loaders.NewRelationLoader(ctx, rq.GetLessonIDsForEpisodes),
+		StudyLessonLinksLoader:    loaders.NewRelationLoader(ctx, rq.GetLinkIDsForLessons),
+		LinkStudyLessonsLoader:    loaders.NewRelationLoader(ctx, rq.GetLessonIDsForLinks),
+	}
+
+	roleLoaders.Store(key, ls)
+
+	return ls
+}
+
+// Will be interesting to see if this will affect anything.
+// I doubt instantiating this 10000 times will be slow,
+// and we'll probably have fewer issues with runaway janitors.
+var profileLoaders = sync.Map{}
+
+func getLoadersForProfile(queries *sqlc.Queries, profileID uuid.UUID) *common.ProfileLoaders {
+	if ls, ok := profileLoaders.Load(profileID); ok {
+		return ls.(*common.ProfileLoaders)
+	}
+
+	ctx := context.TODO()
```

Comment:
Passing a cancelable context into the loaders allows us to control when they terminate later. 

For the moment, I'm letting them live, and later I will cancel the contexts when removing the key from the map.